crates/sui-core/src/authority/consensus_tx_status_cache.rs

```diff
@@ -0,0 +1,316 @@
+    // Transaction is voted to accept by a quorum of validators on fastpath.
+    FastpathCertified,
+    // Transaction is rejected, either by a quorum of validators or indirectly post-commit.
+    Rejected,
+    // Transaction is finalized post commit.
+    Finalized,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum NotifyReadConsensusTxStatusResult {
+    // The consensus position to be read has been updated with a new status.
+    Status(ConsensusTxStatus),
+    // The consensus position to be read has expired.
+    // Provided with the last committed round that was used to check for expiration.
+    Expired(u64),
+}
+
+#[derive(Default)]
+pub(crate) struct ConsensusTxStatusCache {
+    inner: RwLock<Inner>,
+    status_notify_read: NotifyRead<ConsensusTxPosition, ConsensusTxStatus>,
+    /// The depth of the garbage collection.
+    /// We use this to expire positions from old rounds.
+    gc_depth: u64,
+}
+
+#[derive(Default)]
+struct Inner {
+    /// A map of transaction position to its status from consensus.
+    transaction_status: HashMap<ConsensusTxPosition, ConsensusTxStatus>,
+    /// A map of consensus round to all transactions that were updated in that round.
+    round_lookup_map: BTreeMap<u64, HashSet<ConsensusTxPosition>>,
+    /// The last round that was committed, serving as a watermark for expired transactions.
+    last_committed_round: Option<u64>,
+}
+
+impl ConsensusTxStatusCache {
+    pub fn new(gc_depth: u64) -> Self {
+        Self {
+            gc_depth,
+            ..Default::default()
+        }
+    }
+
+    pub fn set_transaction_status(
+        &self,
+        transaction_position: ConsensusTxPosition,
+        status: ConsensusTxStatus,
+    ) {
+        let mut inner = self.inner.write();
+        if let Some(last_committed_round) = inner.last_committed_round {
+            if transaction_position.block.round as u64 + self.gc_depth < last_committed_round {
+                return;
+            }
+        }
+        let old_status = inner
+            .transaction_status
+            .insert(transaction_position, status);
+        if old_status.is_none() {
+            inner
+                .round_lookup_map
+                .entry(transaction_position.block.round as u64)
+                .or_default()
+                .insert(transaction_position);
+        }
+        self.status_notify_read
+            .notify(&transaction_position, &status);
+    }
+
+    pub async fn notify_read_transaction_status(
+        &self,
+        transaction_position: ConsensusTxPosition,
+        old_status: Option<ConsensusTxStatus>,
+    ) -> NotifyReadConsensusTxStatusResult {
+        let registration = self.status_notify_read.register_one(&transaction_position);
+        {
+            let inner = self.inner.read();
+            if let Some(status) = inner.transaction_status.get(&transaction_position) {
+                if Some(status) != old_status.as_ref() {
+                    if let Some(old_status) = old_status {
```

Comment:
How about moving this assert to beginning of the function, to detect the violation earlier?

Reply:
Yeah this is totally minor and optional. I just meant this part seems enforceable before other logic:
```
if let Some(old_status) = old_status {
    assert_eq!(old_status, ConsensusTxStatus::FastpathCertified);
}
```

Reply:
We cannot do this check though.
In the implementation of authority_server, after getting the first status, we always do a tokio select of waiting for a second status and waiting for effects. In this case, if the first status we get is already a finalized, we could end up calling this function with an old status that is not FastpathCertified.

Reply:
Somehow I assumed this is only called when old status is FastpathCertified. I wonder if we should disable the branch waiting for 2nd status when the status is already Finalized. Otherwise I agree the current implementation is the right choice if old status can be any variant.

Reply:
We could. I didn't mostly to avoid having redundant code waiting for effects.